File: backend-ex/controllers/productController.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.singleProduct = exports.removeProduct = exports.listProducts = exports.addProduct = void 0;
const productService_1 = require("../services/productService");
const productModel_1 = require("../models/productModel");
const productImage_1 = require("../models/productImage");
const productSize_1 = require("../models/productSize");
const addProduct = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d;
    try {
        const { name, description, price, category, subCategory, bestseller, sizes, } = req.body;
        const files = req.files;
        if (!((_a = files === null || files === void 0 ? void 0 : files.image1) === null || _a === void 0 ? void 0 : _a[0])) {
            return res.status(400).json({ success: false, message: 'Image is required' });
        }
        const uploadedImages = [(_b = files.image1) === null || _b === void 0 ? void 0 : _b[0], (_c = files.image2) === null || _c === void 0 ? void 0 : _c[0], (_d = files.image3) === null || _d === void 0 ? void 0 : _d[0]].filter(Boolean);
        // Ensure sizes is parsed as array (could be sent as JSON string)
        const parsedSizes = typeof sizes === 'string' ? JSON.parse(sizes) : sizes;
        if (!parsedSizes || !Array.isArray(parsedSizes) || parsedSizes.length === 0) {
            return res.status(400).json({ success: false, message: 'At least one size is required.' });
        }
        const result = yield (0, productService_1.createProductWithAssets)({
            name,
            description,
            price: parseFloat(price),
            category,
            subCategory,
            bestseller: bestseller === 'true' || bestseller === true,
            sizes: parsedSizes,
        }, uploadedImages);
        res.status(201).json({
            success: true,
            productId: result.id,
            message: 'Product created successfully',
        });
    }
    catch (err) {
        console.error(err);
        res.status(500).json({ success: false, message: 'Product creation failed' });
    }
});
exports.addProduct = addProduct;
const listProducts = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const products = yield productModel_1.Product.findAll({
            include: [
                { model: productImage_1.ProductImage, as: 'images' },
                { model: productSize_1.ProductSize, as: 'sizes' },
            ],
        });
        res.json({ success: true, products });
    }
    catch (err) {
        console.error('Error fetching products:', err);
        res.status(500).json({ message: 'Server error fetching products.' });
    }
});
exports.listProducts = listProducts;
const removeProduct = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const productId = Number(req.body.id); // or req.params.id if you prefer
        if (!productId || isNaN(productId)) {
            return res.status(400).json({ message: 'Invalid product ID' });
        }
        yield (0, productService_1.deleteProductById)(productId);
        res.json({ message: 'Product deleted successfully' });
    }
    catch (error) {
        console.error(error);
        res.status(500).json({ message: error.message || 'Server error' });
    }
});
exports.removeProduct = removeProduct;
const singleProduct = (req, res) => __awaiter(void 0, void 0, void 0, function* () {
    try {
        const productId = Number(req.body.id); // or req.params.id if you want to pass id in URL
        if (!productId || isNaN(productId)) {
            return res.status(400).json({ message: 'Invalid product ID' });
        }
        const product = yield productModel_1.Product.findByPk(productId, {
            include: [
                { model: productImage_1.ProductImage, as: 'images' },
                { model: productSize_1.ProductSize, as: 'sizes' },
            ],
        });
        if (!product) {
            return res.status(404);
        }
        res.json(product);
    }
    catch (error) {
        console.error('Error fetching product:', error);
        res.status(500).json({ message: 'Server error fetching product' });
    }
});
exports.singleProduct = singleProduct;

File: backend-go/controllers/product_controller.go
package controllers

import (
	"fmt"
	"mime/multipart"
	"net/http"
	"strconv"

	"github.com/avalent1/snapshop/models"
	"github.com/avalent1/snapshop/services"
	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

type ProductInput struct {
	Name        string   `form:"name" binding:"required" json:"name"`
	Description string   `form:"description" json:"description"`
	Price       float64  `form:"price" binding:"required" json:"price"`
	Category    string   `form:"category" json:"category"`
	SubCategory string   `form:"subCategory" json:"subCategory"`
	Bestseller  bool     `form:"bestseller" json:"bestseller"`
	Sizes       []string `form:"sizes[]" json:"Sizes"`
}

func AddProduct(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		var input ProductInput

		// Parse multipart form
		if err := c.ShouldBind(&input); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"success": false, "message": "Invalid input", "error": err.Error()})
			return
		}

		// Handle images (image1, image2, image3)
		var images []*multipart.FileHeader
		for _, field := range []string{"image1", "image2", "image3"} {
			file, err := c.FormFile(field)
			if err == nil && file != nil {
				images = append(images, file)
			}
		}

		if len(images) == 0 {
			c.JSON(http.StatusBadRequest, gin.H{"success": false, "message": "At least one image is required"})
			return
		}

		// Call service to create product
		product, err := services.CreateProductWithAssets(db, services.NewProductDTO{
			Name:        input.Name,
			Description: input.Description,
			Price:       input.Price,
			Category:    input.Category,
			SubCategory: input.SubCategory,
			Bestseller:  input.Bestseller,
			Sizes:       input.Sizes,
		}, images)

		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"success": false, "message": "Product creation failed", "error": err.Error()})
			return
		}

		c.JSON(http.StatusCreated, gin.H{
			"success":   true,
			"productId": product.ID,
			"message":   "Product created successfully",
		})
	}
}

func ListProducts(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		var products []models.Product

		if err := db.Preload("Images").Preload("Sizes").Find(&products).Error; err != nil {
			fmt.Println("DB Error:", err.Error())
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error fetching products"})
			return
		}

		// Force initialize empty slices to avoid nil on frontend
		for i := range products {
			if products[i].Images == nil {
				products[i].Images = []models.ProductImage{}
			}
			if products[i].Sizes == nil {
				products[i].Sizes = []models.ProductSize{}
			}
		}

		c.JSON(http.StatusOK, gin.H{"success": true, "products": products})
	}
}

func RemoveProduct(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		var requestBody struct {
			ID int `json:"id" binding:"required"`
		}

		// Bind JSON from request body
		if err := c.ShouldBindJSON(&requestBody); err != nil || requestBody.ID == 0 {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid product ID"})
			return
		}

		// Delete the product by ID (implement this service function)
		err := services.DeleteProductByID(db, requestBody.ID)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
			return
		}

		c.JSON(http.StatusOK, gin.H{"message": "Product deleted successfully"})
	}
}

func SingleProduct(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		idStr := c.Param("id")
		id, err := strconv.Atoi(idStr)
		if err != nil || id <= 0 {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Invalid product ID"})
			return
		}

		var product models.Product
		if err := db.Preload("Images").Preload("Sizes").First(&product, id).Error; err != nil {
			if err == gorm.ErrRecordNotFound {
				c.JSON(http.StatusNotFound, gin.H{"message": "Product not found"})
			} else {
				c.JSON(http.StatusInternalServerError, gin.H{"message": "Server error fetching product"})
			}
			return
		}

		c.JSON(http.StatusOK, product)
	}
}
